2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+---
+RFC: RFCxxxx
+Author: Steve Lee
+Status: Draft
+SupercededBy: N/A
+Version: 0.1
+Area: PowerShellGet
+Comments Due: 7/30
+Plan to implement: Yes
+---
+
+# PowerShellGet 3.0 Module
+
+PowerShellGet has enabled an active community to publish and install PowerShell
+Modules.
+However, it has some architectural and design limitations that prevent it from
+moving forward more quickly.
+This RFC proposes some significant changes to address this.
+
+> [!NOTE]
+> There are no intentions to break existing PSGet v2 users so existing
+> users can stay on PSGet v2 and that will continue to work.
+
+## Motivation
+
+    As a PowerShell user,
+    I can discover how to install missing cmdlets,
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Get-Satisfaction
+Get-Satisfaction : The term 'Get-Satisfaction' is not recognized as the name of a cmdlet, function, script file, or operable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+At line:1 char:1
++ Get-Satisfaction
++ ~~~
++ CategoryInfo          : ObjectNotFound: (Get-Satisfaction:String) [], CommandNotFoundException
++ FullyQualifiedErrorId : CommandNotFoundException
+
+Suggestion: You can obtain `Get-Satisfaction` by running `Install-PSResource HappyDays`.
+
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
+In addition, remove dependency on PackageManagement completely as well as dependency on
+nuget.exe.
+This module will only use REST APIs to get and publish nupkgs.
+This module would be shipped in PowerShell 7.
+
+### Side-by-side with PowerShellGet
+
+To ease transition due to the enormity of the breaking change, this module will
+have a different names for the cmdlets to allow for side by side use
+with the existing PowerShellGet module.
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
+
+### Cmdlet compatibility
+
+Function wrappers will be provided to provide compatibility with the two most often
+used cmdlets: `Install-Module` and `Find-Module`.
+These cmdlets will output a warning to use the new cmdlets.
```

Comment:
Why warn?  I'm lukewarm on the name `Install-PSResource` because A) I've just gotten various colleagues trained on using `Install-Module/Update-Module` and B) between the various types of PS resources you list below, IMO the most commonly installed are modules - by a large margin - again IMO.  I'm not sure it is worth "watering down" the noun for module management to address these other, less used resource types.

Reply:
I personally agree with the preference for `Install-Module` etc.

I think the problem that needs to be resolved is:

> How do we keep the command names people know but avoid breaking PowerShellGet v2 side-by-side installation?

Two solutions occur to me:
1. This is the intent of the module path, and shipping PSGet 3 with PS 7 means it comes at the end of the module path by default and PSGet 2 is easy to install ahead of it
2. Perhaps the default cmdlets have a different name but the module exports aliases or functions by default that redirect from the old names. Then there's a way to opt out of this (with a variable or something...).

Reply:
I'm open to just having `Install-Module` and `Find-Module` included forever as I understand that:

1. automation depend on it today
2. muscle memory
3. lots of existing documentation use it

Reply:
The risk I think is like with `Item` -- it's not clear what a `PSResource` refers to compared to a `Module`. It makes lots of sense that the PSGet can handle more than just modules, but providing a concrete conceptual schema prevents the module noun becoming a second order simulacrum of sorts

Reply:
I don't have a solution, but I think it's problematic to have 2 cmdlets in different modules with the same name which are mostly but not entirely compatible.

Reply:
Expectation is that `Find-Module` and `Install-Module` is completely compatible with the v2 counterparts.  Only the internals are changing.

Reply:
Will the output object contain the same properties?

Reply:
So `Install-PSResource` will support, in it's "RequiredModules" (err, resources?) file or parameter the nuget syntax: `[1.0,2.0)` meaning 1.0 or greater but not 2.0 ... and so on, including `1.*` or even `(1.1, 2.0)` ....

But Install-Module will still use the confusingly named `-ModuleVersion` and the inclusive-only `-MaximumVersion` and you're going to map these with the weird 1.99999999.9999999 semantics they're doing in PowerShellGet now?